NixOS/nixpkgs
# n8n: 1.9.3 -> 1.46.0, repackage with pnpm.fetchDeps

Since https://github.com/NixOS/nixpkgs/pull/290715 was merged, it's much easier to package and maintain pnpm projects.

## Description of changes

- replace all the `node-env`, `node-composition` stuff with `pnpm.fetchDeps` and `pnpm.configHook`
- upgrade n8n to latest version
- move to pkgs/by-name
- set N8N_RELEASE_TYPE to stable
- remove old patches and build inputs that seem no longer necessary
  - `oracledb` and `@sap/hana-client` removal on aarch64: these don't exist in node_modules anymore and aarch64 build works
  - `pkgs.which`: probably was unused before, there were no references to it
  - `NIX_CFLAGS_COMPILE = "-DNODE_API_EXPERIMENTAL_NOGC_ENV_OPT_OUT";`: builds without this with node 20
- remove update script: there's no update script for similarly packaged apps, and updating is fairly simple
- format with nixfmt-rfc-style

## Issues I encountered and solved

### Error: cannot find '/build/source/node_modules/.pnpm/sqlite3@5.1.7/node_modules/sqlite3/build/Release/node_sqlite3.node

`pnpm.configHook` calls `pnpm install` with `--ignore-scripts` that skips downloading/building node-sqlite3 bindings. A solution is building it manually: (use `node-gyp` from nixpkgs rather than from pnpm installed node_modules to avoid other errors)

https://github.com/NixOS/nixpkgs/blob/077adff3243b078a0f4a49421038d6f3f16fdf37/pkgs/by-name/n8/n8n/package.nix#L65-L67

### turbo run build: Failed to create APIClient: Unable to set up TLS. builder error: No such file or directory

In turbo release 1.12.4, they switched to using `rustls-tls-native-roots` rust crate which tries to open a file from `SSL_CERT_FILE` environment variable, which is set to `/no-cert-file.crt` in a nix build environment. Adding `cacert` to build inputs will set this env var to a real file, fixing the issue.

### ERR_PNPM_BAD_PM_VERSION This project is configured to use v9.1.4 of pnpm. Your current pnpm is v9.1.1.

Patching `package.json` to replace pnpm version to the current fixes it:

https://github.com/NixOS/nixpkgs/blob/077adff3243b078a0f4a49421038d6f3f16fdf37/pkgs/by-name/n8/n8n/package.nix#L35



## Things done



- Built on platform(s)
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)



---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc
